### DISCUSSION TITLE
How to write grammar rule?

### DISCUSSION THREAD

**Discussion by dprogm** [pest-parser/pest#883] [General]
Hello everybody :wink:

I have perhaps a simple question but cannot answer it by myself. Suppose the following grammar rules:

```
managed_interfaces = { "manages" ~ (ws+ ~ (interface | fqn)) ~ (ws* ~ "," ~ ws* ~ (interface | fqn))* }

interface = { "interface" ~ ws+ ~ id ~ (ws+ ~ "extends" ~ (ws+ ~ (interface | fqn)))? ~ managed_interfaces? ~ "{” }
```

The problem is that I want to require whitespace before `managed_interfaces` if the last group ends with an `fqn` or is not present at all and optional whitespace in case of an `interface`. I don't know exactly how to express this in a clean and readable way.

Any tip is appreciated :relaxed:

**Comment by tomtau** [comment]
Maybe splitting the `interface` rule into two rules: one for when it ends with `fqn` or is not present at all (with mandatory whitespace), and another for when it ends with `interface` (with optional whitespace)?
